### PR TITLE
Pysp persistent enhancements

### PR DESCRIPTION
**Pull request by jwatsonnm**
## Fixes

The main bug addressed by this PR relates to the use of PH solver servers and persistent interfaces, in the context when multiple scenarios/bundles were associated with a PH solver server. Because there was a single solver plugin per PH solver server, multiple instances were bound to the same solver plugin. Specifically, the last instances compiled to the (single) persistent solver plugin was "the" instance for the PH solver server. Leading to, well, undefined / random behaviors. Fortunately, almost nobody ran in this mode, and nobody ran in this mode outside experimentation. This PR also fully generalizes the persistent interfaces for PH, specifically for bundles and serial execution mode.

## Summary/Motivation:

Primary purpose of this PR is to address some existing issues with PH and persistent solver interfaces, and to generalize PH to work "maximally" (whatever that means) with the new persistent solver interfaces. The main generalization is moving from a single solver plugin on the PH base class to a map of solver plugins, one for each object (scenario or bundle) under consideration. This corrects numerous major bugs with distributed PH in the context of multiple scenarios/bundles per phsolverserver process, and allows for use of persistent interfaces in serial PH. 

## Changes proposed in this PR:
- Generalization of solver plugin to solver plugin map for PH base object - see above.
- Minor cleanup of reporting by Eckstein-Combettes extension.
- Added checks to PH solver servers to ensure that when using persistent solver interfaces, the instances the server is asked to solve are actually the ones that it has been configured for.
- Added full bundle processing capability for persistent solver interfaces - this did not previously exist. 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/pysp/plugins/ecksteincombettesextension.py [hunk 1]
```diff
@@ -363,11 +364,12 @@ def compute_updates(self, ph, subproblems, scenario_solve_counts):
 
         print("Computed sub-phi values (scenario, phi, iters-since-last-incorporated):")
         for sub_phi in sorted(sub_phi_to_scenario_map.keys()):
-            print("  %16e: " % sub_phi),
+            print("  %16e: " % sub_phi, end="")
```
Comment (ghackebeil) [inline pyomo/pysp/plugins/ecksteincombettesextension.py]: This is not going to work for Python 2. You can probably use `six.print_` instead.

File: pyomo/pysp/plugins/ecksteincombettesextension.py [hunk 2]
```diff
@@ -394,9 +396,8 @@ def compute_updates(self, ph, subproblems, scenario_solve_counts):
             for phi in sorted_phis[0:ph._async_buffer_length]:
                 if ((self._queue_only_negative_subphi_subproblems) and (phi < 0.0)) or (not self._queue_only_negative_subphi_subproblems):
                     scenario_name = sub_phi_to_scenario_map[phi][0] 
-                    print("%30s %16e" % (scenario_name,phi)),
+                    print("%30s %16e" % (scenario_name,phi), end="")
```
Comment (ghackebeil) [inline pyomo/pysp/plugins/ecksteincombettesextension.py]: same as above

File: pyomo/pysp/plugins/ecksteincombettesextension.py (hunk 1 above)
Comment (jwatsonnm) [reply pyomo/pysp/plugins/ecksteincombettesextension.py]: Thanks for pointing this out - I've checked to use six.print_.

File: pyomo/pysp/plugins/ecksteincombettesextension.py (hunk 2 above)
Comment (jwatsonnm) [reply pyomo/pysp/plugins/ecksteincombettesextension.py]: See above - fixed!

**Comment by jwatsonnm** [comment]
Gabe's changes look good to me - I examined the diffs in some detail. If tests look good, I will merge and close tomorrow AM.

**Pull request merged**